rustls/tests/unbuffered.rs

```diff
@@ -0,0 +1,204 @@
+#![cfg(any(feature = "ring", feature = "aws_lc_rs"))]
+use std::sync::Arc;
+
+use rustls::client::{ClientConnectionData, UnbufferedClientConnection};
+use rustls::server::{ServerConnectionData, UnbufferedServerConnection};
+use rustls::unbuffered::{ConnectionState, UnbufferedConnectionCommon, UnbufferedStatus};
+
+use crate::common::*;
+
+mod common;
+
+#[test]
+fn handshake() {
+    for version in rustls::ALL_VERSIONS {
+        let server_config = make_server_config(KeyType::Rsa);
+        let client_config = make_client_config_with_versions(KeyType::Rsa, &[version]);
+
+        let mut client =
+            UnbufferedClientConnection::new(Arc::new(client_config), server_name("localhost"))
+                .unwrap();
+        let mut server = UnbufferedServerConnection::new(Arc::new(server_config)).unwrap();
+        let mut buffers = BothBuffers::default();
+
+        let mut count = 0;
+        let mut client_handshake_done = false;
+        let mut server_handshake_done = false;
+        while !client_handshake_done || !server_handshake_done {
+            match advance_client(&mut client, &mut buffers.client, NO_ACTIONS) {
+                State::EncodedTlsData => {}
+                State::MustTransmitTlsData => buffers.client_send(),
+                State::NeedsMoreTlsData => buffers.server_send(),
+                State::TrafficTransit => client_handshake_done = true,
+            }
+
+            match advance_server(&mut server, &mut buffers.server, NO_ACTIONS) {
+                State::EncodedTlsData => {}
+                State::MustTransmitTlsData => buffers.server_send(),
+                State::NeedsMoreTlsData => buffers.client_send(),
+                State::TrafficTransit => server_handshake_done = true,
+            }
```

Comment:
At the moment this test is more like example code, and doesn't fail unless it can't be completed within 100 events.

It might be better for this test to observe the precise expected sequences of the underlying `ConnectionState` events (though not the values or lengths of data involved).

Reply:
An idea to do this:

- have `let mut client_transcript = Vec::<String>::new();`, ditto for server
- for every each `ConnectionState` seen by `advance_client`, push the output of impl Debug of  it (ditto server) -- this also will provide coverage of that code
- at the end of the test, `assert_eq!(client_transcript, vec![ "MustEncodeTlsData", "MustTransmitTlsData", ... ]);`